File: parser/data_parser.py
# -*- coding: utf-8 -*-
"""
Parse extracted PDF text to find staff assignments.
Matches names against the master staff list.
"""

import re
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from staff_data import get_staff_dict, get_all_names


def normalize_name(name):
    """Normalize Thai name: remove extra spaces, normalize sara am, etc."""
    # Remove zero-width characters and normalize whitespace
    name = re.sub(r'[\u200b\u200c\u200d\ufeff]', '', name)
    # Collapse multiple spaces to one
    name = re.sub(r'\s+', ' ', name).strip()
    # Remove trailing ์ that got separated by space (common PDF artifact)
    name = re.sub(r'\s+์', '์', name)
    # Remove trailing ิ that got separated
    name = re.sub(r'\s+ิ', 'ิ', name)
    # Remove trailing ี that got separated
    name = re.sub(r'\s+ี', 'ี', name)
    # Remove trailing ื that got separated
    name = re.sub(r'\s+ื', 'ื', name)
    # Remove trailing ่ that got separated
    name = re.sub(r'\s+่', '่', name)
    # Remove trailing ้ that got separated
    name = re.sub(r'\s+้', '้', name)
    # Remove trailing ๊ that got separated
    name = re.sub(r'\s+๊', '๊', name)
    # Remove trailing ็ that got separated
    name = re.sub(r'\s+็', '็', name)
    return name


def build_name_lookup(staff_dict):
    """
    Build lookup structures for fuzzy name matching.
    Returns dict mapping normalized_name -> original_name
    """
    lookup = {}
    for name in staff_dict:
        norm = normalize_name(name)
        lookup[norm] = name
        # Also index by last name for partial matching
        parts = name.split()
        if len(parts) >= 2:
            lookup[parts[-1]] = name  # last name only (may collide, but helps)
    return lookup


def find_matching_staff(extracted_name, name_lookup, staff_dict):
    """
    Try to match an extracted name against the staff list.
    Returns the canonical staff name or None.
    """
    norm = normalize_name(extracted_name)

    # Direct match
    if norm in staff_dict:
        return norm

    # Normalized match
    if norm in name_lookup:
        return name_lookup[norm]

    # Fuzzy: try removing all spaces and matching
    no_space = norm.replace(' ', '')
    for staff_name in staff_dict:
        if staff_name.replace(' ', '') == no_space:
            return staff_name

    # Fuzzy: try matching last name + partial first name
    parts = norm.split()
    if len(parts) >= 2:
        last = parts[-1]
        first = parts[0]
        for staff_name in staff_dict:
            sparts = staff_name.split()
            if len(sparts) >= 2:
                # Same last name and first name starts with same chars
                if sparts[-1] == last and (
                    sparts[0].startswith(first[:3]) or first.startswith(sparts[0][:3])
                ):
                    return staff_name

    return None


def extract_order_info(text):
    """Extract order number and subject from the PDF text."""
    order_number = None
    order_subject = None
    order_date = None

    # Match order number: ที่ XXX / 2569 or คำสั่งที่ XXX/2569
    m = re.search(r'(?:ท[ี่ ]+|คำสั่ง(?:ที่)?)\s*(\d+)\s*/\s*(\d{4})', text)
    if m:
        order_number = f"{m.group(1)}/{m.group(2)}"

    # Match subject: เรื่อง ...
    m = re.search(r'เรื่อง\s+(.+?)(?:\n|$)', text)
    if m:
        order_subject = normalize_name(m.group(1).strip())

    # Match date: สั่ง ณ วันที่ ... or วันที่ ...เดือน...พ.ศ. ...
    m = re.search(r'สั่ง\s*ณ\s*วันที่\s*(\d+)\s*เดือน\s*(\S+)\s*พ\.ศ\.\s*(\d+)', text)
    if m:
        order_date = f"{m.group(1)} {m.group(2)} {m.group(3)}"
    else:
        # Try another pattern: วันเสาร์ที่ XX เดือน ...
        m = re.search(r'วัน\S+ที่\s+(\d+)\s+(?:เดือน\s+)?(\S+)\s+(?:พ\.ศ\.\s*)?(\d{4})', text)
        if m:
            order_date = f"{m.group(1)} {m.group(2)} {m.group(3)}"

    return {
        'order_number': order_number,
        'order_subject': order_subject,
        'order_date': order_date,
    }


def extract_section_info(text):
    """Extract section/committee headers and date context."""
    sections = []
    # Match section headers like: ๑. คณะกรรมการอำนวยการ or 4.1 ชั้นมัธยมศึกษาปีที่ 1
    pattern = r'(?:^|\n)\s*(?:[\d๐-๙]+[\.\)]\s*(?:[\d๐-๙]+[\.\)]?\s*)?)(คณะกรรมการ\S+|ฝ่าย\S+|ชั้นมัธยม\S+)'
    for m in re.finditer(pattern, text):
        sections.append(m.group(1))
    return sections


def extract_date_contexts(text):
    """Extract date-duty mappings from the text."""
    dates = []
    # Pattern: วันเสาร์ที่ XX มีนาคม 2569 or วันอาทิตย์ที่ XX เมษายน 2569
    pattern = r'วัน(\S+?)ที่\s*(\d+)\s+(?:เดือน\s*)?(\S+?)\s+(?:พ\.ศ\.\s*)?(\d{4})'
    for m in re.finditer(pattern, text):
        dates.append({
            'day_name': m.group(1),
            'day': m.group(2),
            'month': m.group(3),
            'year': m.group(4),
            'full': f"วัน{m.group(1)}ที่ {m.group(2)} {m.group(3)} {m.group(4)}",
            'pos': m.start(),
        })
    return dates


def parse_assignments(text, source_file=""):
    """
    Main parsing function: extract all staff assignments from PDF text.
    Returns list of dicts with assignment info.
    """
    staff_dict = get_staff_dict()
    name_lookup = build_name_lookup(staff_dict)

    order_info = extract_order_info(text)
    date_contexts = extract_date_contexts(text)

    assignments = []
    found_names = set()

    # Pattern to find Thai names with title prefixes
    name_pattern = re.compile(
        r'(?:นาย|นาง(?:สาว)?|น\.ส\.)\s*'
        r'([\u0E00-\u0E7F\s]+?)(?=\s+(?:'
        r'ผู้อำนวยการ|รองผู้อำนวยการ|หัวหน้า|ประธาน|รองประธาน|กรรมการ|'
        r'ครู|พนักงาน|ปฏิบัติ|มีหน้าที่|'
        r'นาย|นาง|น\.ส\.|$|'
        r'\d+\.|ห้อง|ม\.\s*\d|ระดับ|ตรวจ|รับ'
        r'))',
        re.MULTILINE
    )

    # Also match names in table-like format: น.ส.ชื่อ นามสกุล or นางชื่อ นามสกุล
    name_pattern2 = re.compile(
        r'(?:นาย|นาง(?:สาว)?|น\.ส\.)\s*'
        r'([\u0E00-\u0E7F]+\s+[\u0E00-\u0E7F]+)',
    )

    # Collect all name matches with positions
    all_matches = []
    for m in name_pattern.finditer(text):
        raw = m.group(1).strip()
        if len(raw) > 3:  # minimum reasonable name length
            all_matches.append((m.start(), raw))

    for m in name_pattern2.finditer(text):
        raw = m.group(1).strip()
        if len(raw) > 3:
            # Avoid duplicates at same position
            pos = m.start()
            if not any(abs(pos - p) < 5 for p, _ in all_matches):
                all_matches.append((pos, raw))

    # Sort by position in text
    all_matches.sort(key=lambda x: x[0])

    # Determine date context for each match
    for pos, raw_name in all_matches:
        name = normalize_name(raw_name)

        # Clean up: remove trailing non-name words
        name = re.sub(
            r'\s*(?:ประธาน|รองประธาน|กรรมการ|เลขานุการ|ผู้ช่วย|หัวหน้า|ครู|'
            r'ปฏิบัติ|ผู้อำนวยการ|รองผู้|พนักงาน|ช่วยราชการ).*$',
            '', name
        ).strip()

        if len(name) < 4:
            continue

        matched = find_matching_staff(name, name_lookup, staff_dict)
        if matched:
            found_names.add(matched)

            # Find closest date context
            duty_date = ""
            for dc in date_contexts:
                if dc['pos'] <= pos:
                    duty_date = dc['full']

            # Find section context (look backwards for section header)
            section = ""
            section_pattern = re.compile(
                r'(?:คณะกรรมการ\S+|ฝ่าย\S+|'
                r'ชั้นมัธยมศึกษาปีที่\s*\d|'
                r'ห้องสอบที่\s*\d+|'
                r'ม\.\s*\d+/\d+)',
            )
            text_before = text[max(0, pos - 500):pos]
            section_matches = list(section_pattern.finditer(text_before))
            if section_matches:
                section = section_matches[-1].group(0)

            staff_info = staff_dict[matched]
            assignments.append({
                'name': matched,
                'department': staff_info['department'],
                'position': staff_info['position'],
                'admin_role': staff_info['admin_role'],
                'order_number': order_info['order_number'] or '',
                'order_subject': order_info['order_subject'] or '',
                'order_date': order_info['order_date'] or '',
                'duty_date': duty_date,
                'duty_section': section,
                'source_file': source_file,
            })

    return {
        'order_info': order_info,
        'assignments': assignments,
        'unique_staff': list(found_names),
        'total_assignments': len(assignments),
        'unique_count': len(found_names),
    }

File: js/data-parser.js
// Parse extracted PDF text to find staff assignments
// Port of parser/data_parser.py

import { getStaffDict, getAllNames } from './staff-data.js';

function normalizeName(name) {
  name = name.replace(/[\u200b\u200c\u200d\ufeff]/g, '');
  name = name.replace(/\s+/g, ' ').trim();
  name = name.replace(/\s+์/g, '์');
  name = name.replace(/\s+ิ/g, 'ิ');
  name = name.replace(/\s+ี/g, 'ี');
  name = name.replace(/\s+ื/g, 'ื');
  name = name.replace(/\s+่/g, '่');
  name = name.replace(/\s+้/g, '้');
  name = name.replace(/\s+๊/g, '๊');
  name = name.replace(/\s+็/g, '็');
  return name;
}

function buildNameLookup(staffDict) {
  const lookup = {};
  for (const name of Object.keys(staffDict)) {
    const norm = normalizeName(name);
    lookup[norm] = name;
    const parts = name.split(/\s+/);
    if (parts.length >= 2) lookup[parts[parts.length - 1]] = name;
  }
  return lookup;
}

function findMatchingStaff(extractedName, nameLookup, staffDict) {
  const norm = normalizeName(extractedName);

  if (staffDict[norm]) return norm;
  if (nameLookup[norm]) return nameLookup[norm];

  const noSpace = norm.replace(/\s/g, '');
  for (const staffName of Object.keys(staffDict)) {
    if (staffName.replace(/\s/g, '') === noSpace) return staffName;
  }

  const parts = norm.split(/\s+/);
  if (parts.length >= 2) {
    const last = parts[parts.length - 1];
    const first = parts[0];
    for (const staffName of Object.keys(staffDict)) {
      const sparts = staffName.split(/\s+/);
      if (sparts.length >= 2) {
        if (sparts[sparts.length - 1] === last &&
            (sparts[0].startsWith(first.substring(0, 3)) || first.startsWith(sparts[0].substring(0, 3)))) {
          return staffName;
        }
      }
    }
  }

  return null;
}

function extractOrderInfo(text) {
  let orderNumber = null, orderSubject = null, orderDate = null;

  let m = text.match(/(?:ท[ี่ ]+|คำสั่ง(?:ที่)?)\s*(\d+)\s*\/\s*(\d{4})/);
  if (m) orderNumber = `${m[1]}/${m[2]}`;

  m = text.match(/เรื่อง\s+(.+?)(?:\n|$)/);
  if (m) orderSubject = normalizeName(m[1].trim());

  m = text.match(/สั่ง\s*ณ\s*วันที่\s*(\d+)\s*เดือน\s*(\S+)\s*พ\.ศ\.\s*(\d+)/);
  if (m) {
    orderDate = `${m[1]} ${m[2]} ${m[3]}`;
  } else {
    m = text.match(/วัน\S+ที่\s+(\d+)\s+(?:เดือน\s+)?(\S+)\s+(?:พ\.ศ\.\s*)?(\d{4})/);
    if (m) orderDate = `${m[1]} ${m[2]} ${m[3]}`;
  }

  return { order_number: orderNumber, order_subject: orderSubject, order_date: orderDate };
}

function extractDateContexts(text) {
  const dates = [];
  const re = /วัน(\S+?)ที่\s*(\d+)\s+(?:เดือน\s*)?(\S+?)\s+(?:พ\.ศ\.\s*)?(\d{4})/g;
  let m;
  while ((m = re.exec(text)) !== null) {
    dates.push({
      day_name: m[1], day: m[2], month: m[3], year: m[4],
      full: `วัน${m[1]}ที่ ${m[2]} ${m[3]} ${m[4]}`,
      pos: m.index,
    });
  }
  return dates;
}

function extractTimeContexts(text) {
  const contexts = [];
  const re = /(ภาคเช้า|ภาคบ่าย)?\s*เวลา\s*(\d{1,2})[.:](\d{2})\s*น\.?\s*[-–]\s*(\d{1,2})[.:](\d{2})\s*น\.?/g;
  let m;
  while ((m = re.exec(text)) !== null) {
    const period = m[1] || '';
    const h1 = parseInt(m[2]), m1 = parseInt(m[3]);
    const h2 = parseInt(m[4]), m2 = parseInt(m[5]);
    const startMin = h1 * 60 + m1;
    const endMin = h2 * 60 + m2;
    const duration = Math.max(endMin - startMin, 0) / 60.0;
    let timeStr = `${String(h1).padStart(2,'0')}:${String(m1).padStart(2,'0')}-${String(h2).padStart(2,'0')}:${String(m2).padStart(2,'0')}`;
    if (period) timeStr = `${period} ${timeStr}`;
    contexts.push({
      period, start_hour: h1, start_min: m1, end_hour: h2, end_min: m2,
      duration_hours: Math.round(duration * 100) / 100,
      time_str: timeStr, pos: m.index,
    });
  }
  return contexts;
}

export function parseAssignments(text, sourceFile = '', staffDictOverride = null) {
  const staffDict = staffDictOverride || getStaffDict();
  const nameLookup = buildNameLookup(staffDict);
  const orderInfo = extractOrderInfo(text);
  const dateContexts = extractDateContexts(text);
  const timeContexts = extractTimeContexts(text);

  const assignments = [];
  const foundNames = new Set();

  // Pattern 1: Thai names with title prefixes
  const namePattern1 = /(?:นาย|นาง(?:สาว)?|น\.ส\.)\s*([\u0E00-\u0E7F\s]+?)(?=\s+(?:ผู้อำนวยการ|รองผู้อำนวยการ|หัวหน้า|ประธาน|รองประธาน|กรรมการ|ครู|พนักงาน|ปฏิบัติ|มีหน้าที่|นาย|นาง|น\.ส\.|$|\d+\.|ห้อง|ม\.\s*\d|ระดับ|ตรวจ|รับ))/gm;
  // Pattern 2: Table-like format
  const namePattern2 = /(?:นาย|นาง(?:สาว)?|น\.ส\.)\s*([\u0E00-\u0E7F]+\s+[\u0E00-\u0E7F]+)/g;

  const allMatches = [];

  let m;
  while ((m = namePattern1.exec(text)) !== null) {
    const raw = m[1].trim();
    if (raw.length > 3) allMatches.push([m.index, raw]);
  }

  while ((m = namePattern2.exec(text)) !== null) {
    const raw = m[1].trim();
    if (raw.length > 3) {
      const pos = m.index;
      if (!allMatches.some(([p]) => Math.abs(pos - p) < 5)) {
        allMatches.push([pos, raw]);
      }
    }
  }

  allMatches.sort((a, b) => a[0] - b[0]);

  for (const [pos, rawName] of allMatches) {
    let name = normalizeName(rawName);
    name = name.replace(/\s*(?:ประธาน|รองประธาน|กรรมการ|เลขานุการ|ผู้ช่วย|หัวหน้า|ครู|ปฏิบัติ|ผู้อำนวยการ|รองผู้|พนักงาน|ช่วยราชการ).*$/, '').trim();

    if (name.length < 4) continue;

    const matched = findMatchingStaff(name, nameLookup, staffDict);
    if (!matched) continue;

    foundNames.add(matched);

    let dutyDate = '';
    for (const dc of dateContexts) {
      if (dc.pos <= pos) dutyDate = dc.full;
    }

    let dutyTime = '', durationHours = 0;
    for (const tc of timeContexts) {
      if (tc.pos <= pos) {
        dutyTime = tc.time_str;
        durationHours = tc.duration_hours;
      }
    }

    let section = '';
    const sectionRe = /(?:คณะกรรมการ\S+|ฝ่าย\S+|ชั้นมัธยมศึกษาปีที่\s*\d|ห้องสอบที่\s*\d+|ม\.\s*\d+\/\d+)/g;
    const textBefore = text.substring(Math.max(0, pos - 500), pos);
    let sm;
    while ((sm = sectionRe.exec(textBefore)) !== null) section = sm[0];

    const staffInfo = staffDict[matched];
    assignments.push({
      name: matched,
      department: staffInfo.department,
      position: staffInfo.position,
      admin_role: staffInfo.admin_role,
      order_number: orderInfo.order_number || '',
      order_subject: orderInfo.order_subject || '',
      order_date: orderInfo.order_date || '',
      duty_date: dutyDate,
      duty_time: dutyTime,
      duration_hours: durationHours,
      duty_section: section,
      source_file: sourceFile,
    });
  }

  return {
    order_info: orderInfo,
    assignments,
    unique_staff: [...foundNames],
    total_assignments: assignments.length,
    unique_count: foundNames.size,
  };
}
